File: src/logic/duels.js
// Logique des duels : charge fixe, le plus de répétitions gagne.
// Format en 3 rounds — premier à 2 victoires :
//   Round 1 : le CHALLENGER (celui qui lance le duel) choisit l'exercice.
//   Round 2 : l'ADVERSAIRE choisit l'exercice.
//   Round 3 : l'IA choisit l'exercice de départage.
import { baremes } from '../data/clubSP';

// Compte les rounds gagnés par chacun.
export function comptageVictoires(duel) {
  let moi = 0;
  let lui = 0;
  duel.rounds.forEach((round) => {
    if (round.mesReps == null || round.sesReps == null) return; // pas encore joué
    if (round.mesReps > round.sesReps) moi++;
    else if (round.sesReps > round.mesReps) lui++;
    // Égalité de reps : personne ne marque, le round sera rejoué.
  });
  return { moi, lui };
}

// Statut d'un duel à partir de ses rounds.
export function statutDuel(duel) {
  const { moi, lui } = comptageVictoires(duel);
  if (moi >= 2) return 'gagné';
  if (lui >= 2) return 'perdu';
  return 'en cours';
}

// Joue le round de départage : l'IA choisit un exercice en kg au hasard,
// avec une charge fixe (le palier Bronze de l'exercice, accessible aux deux).
// Les répétitions sont SIMULÉES pour l'instant — le vrai duel viendra avec le backend.
// L'IA choisit un exercice en kg au hasard, charge fixe = palier Bronze (accessible aux deux).
export function exerciceAleatoireIA(sexe) {
  const exercicesKg = Object.entries(baremes[sexe]).filter(([, b]) => b.unite === 'kg');
  const [exercice, bareme] = exercicesKg[Math.floor(Math.random() * exercicesKg.length)];
  return { exercice, charge: bareme.paliers[0] };
}

export function jouerRoundIA(duel, sexe) {
  const { exercice, charge } = exerciceAleatoireIA(sexe);
  const rounds = duel.rounds.map((round) =>
    round.exercice === null
      ? {
          ...round,
          exercice,
          charge,
          mesReps: 6 + Math.floor(Math.random() * 10),
          sesReps: 6 + Math.floor(Math.random() * 10),
        }
      : round
  );
  const nouveau = { ...duel, rounds };
  nouveau.statut = statutDuel(nouveau);
  return nouveau;
}

File: backend/app/duels.py
"""Logique des duels BO3 — version Python (miroir de src/logic/duels.js).

Règles (identiques au front) :
- Charge fixe, le plus de répétitions gagne le round.
- Premier à 2 rounds gagnés remporte le duel (BO3 = best of 3).
- Round 1 : exercice choisi par le CHALLENGER (celui qui lance le duel).
- Round 2 : exercice choisi par l'ADVERSAIRE.
- Round 3 (départage) : exercice choisi par l'IA, au hasard parmi les
  exercices en kg, charge fixe = palier Bronze (accessible aux deux).
- Égalité de reps sur un round : personne ne marque, le round est rejoué.
"""

import random
import secrets

from .baremes import BAREMES

# Alphabet du code de duel : sans 0/O ni 1/I/L (trop ambigus à l'oral/à l'écrit).
ALPHABET_CODE = "ABCDEFGHJKMNPQRSTUVWXYZ23456789"


def generer_code(longueur: int = 6) -> str:
    """Code court à partager pour rejoindre un duel en ligne (ex. 'K7XPQR')."""
    return "".join(secrets.choice(ALPHABET_CODE) for _ in range(longueur))


def comptage_victoires(rounds: list) -> dict:
    """Compte les rounds gagnés par le challenger et par l'adversaire.

    Un round = {"reps_challenger": int|None, "reps_adversaire": int|None, ...}.
    Un round pas encore joué (reps à None) ne compte pas.
    """
    challenger = 0
    adversaire = 0
    for round_ in rounds:
        reps_c = round_.get("reps_challenger")
        reps_a = round_.get("reps_adversaire")
        if reps_c is None or reps_a is None:
            continue  # pas encore joué
        if reps_c > reps_a:
            challenger += 1
        elif reps_a > reps_c:
            adversaire += 1
        # Égalité : personne ne marque, le round sera rejoué.
    return {"challenger": challenger, "adversaire": adversaire}


def gagnant_duel(rounds: list) -> str | None:
    """Renvoie 'challenger' ou 'adversaire' si l'un a 2 victoires, sinon None."""
    score = comptage_victoires(rounds)
    if score["challenger"] >= 2:
        return "challenger"
    if score["adversaire"] >= 2:
        return "adversaire"
    return None


def exercice_aleatoire_ia(sexe: str) -> dict:
    """L'IA choisit un exercice en kg au hasard, charge = palier Bronze.

    (Le palier Bronze est accessible aux deux joueurs — duel équitable.)
    """
    exercices_kg = [
        (nom, bareme)
        for nom, bareme in BAREMES[sexe].items()
        if bareme["unite"] == "kg"
    ]
    nom, bareme = random.choice(exercices_kg)
    return {"exercice": nom, "charge": bareme["paliers"][0]}


def round_verrouille(round_: dict) -> bool:
    """Vrai si le round a un résultat DÉCISIF (reps différentes) — verrouillé,
    on ne peut plus le rejouer. Une égalité (mêmes reps) reste rejouable
    (le round « sera rejoué », voir comptage_victoires)."""
    reps_c, reps_a = round_.get("reps_challenger"), round_.get("reps_adversaire")
    return reps_c is not None and reps_a is not None and reps_c != reps_a
